services/web/server/src/simcore_service_webserver/projects/_permalink.py

```diff
@@ -0,0 +1,87 @@
+import asyncio
+import logging
+from typing import Any, Callable, Coroutine
+
+from aiohttp import web
+from models_library.projects import ProjectID
+from pydantic import BaseModel, HttpUrl
+from simcore_service_webserver.projects.project_models import ProjectDict
+
+from .projects_exceptions import PermalinkFactoryError, PermalinkNotAllowedError
+
+_PROJECT_PERMALINK = f"{__name__}"
+_logger = logging.getLogger(__name__)
+
+
+class ProjectPermalink(BaseModel):
+    url: HttpUrl
+    is_public: bool
+
+
+_CreateLinkCallable = Callable[
+    [web.Request, ProjectID], Coroutine[Any, Any, ProjectPermalink]
+]
+
+
+def register_factory(app: web.Application, factory_coro: _CreateLinkCallable):
+    if _create := app.get(_PROJECT_PERMALINK):
+        raise PermalinkFactoryError(
+            f"Permalink factory can only be set once: registered {_create}"
+        )
+    app[_PROJECT_PERMALINK] = factory_coro
+
+
+def _get_factory(app: web.Application) -> _CreateLinkCallable:
+    if _create := app.get(_PROJECT_PERMALINK):
+        return _create
+
+    raise PermalinkFactoryError(
+        "Undefined permalink factory. Check plugin initialization."
+    )
+
+
+_CALLBACK_TIMEOUT = (
+    2  #  get_plugin_settings(request.app).PROJECTS_PERMALINK_FACTORY_TIMEOUT
+)
+
+
+async def _create_permalink(
+    request: web.Request, project_id: ProjectID
+) -> ProjectPermalink:
+
+    create = _get_factory(request.app)
+    assert create  # nosec
+
+    try:
+        permalink: ProjectPermalink = await asyncio.wait_for(
+            create(request, project_id), timeout=_CALLBACK_TIMEOUT
+        )
+        return permalink
+    except asyncio.TimeoutError as err:
+        raise PermalinkFactoryError(
+            f"Permalink factory callback '{create}' timed out after {_CALLBACK_TIMEOUT} secs"
+        ) from err
+
+
+async def update_or_pop_permalink_in_project(
+    request: web.Request, project: ProjectDict
+) -> ProjectPermalink | None:
+    """Updates permalink entry in project
+
+    Creates a permalink for this project and assigns it to project["permalink"]
+
+    If fails, it pops it from project (so it is not set in the pydantic model. SEE ProjectGet.permalink)
```

Comment:
This is related with my comment above. The idea is that this function provides some guarantees on the state of ``projects['permalink']`` given a request.
Therefore it will update or delete (i.e. pop) it form the ``projects`` dict